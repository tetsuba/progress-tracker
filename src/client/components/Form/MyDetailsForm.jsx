import React, { useState } from 'react'
import { Button, Form, Row } from 'react-bootstrap'
import Box from '../Box/Box'
import { useInputChange } from '../../hooks/hooks'
import { useMutation } from '@apollo/react-hooks'
import { UPDATE_USER_DETAILS_MUTATION } from '../../api/user/user.mutation'
import { GET_USER_DETAILS_QUERY } from '../../api/user/user.query'

// TYPES
import type { userDetailsType } from '../../types/userType'

type Props = {
  user: userDetailsType,
}

export default function MyDetailsForm(props: Props) {
  const { user } = props
  const [updateUserDetails] = useMutation(UPDATE_USER_DETAILS_MUTATION, {
    refetchQueries: [{ query: GET_USER_DETAILS_QUERY }],
  })
  const [errorMessage, setErrorMessage] = useState('')
  const [inputs, setInputs] = useInputChange({
    firstName: user.firstName,
    lastName: user.lastName,
  })
  const [showFormElement, setShowFormElement] = useState(false)
  const options = {
    variables: {
      input: {
        ...inputs,
      },
    },
  }

  return (
    <Row>
      <Box max={500}>
        <h3 className="mb-5">My Details</h3>
        <Form
          id="MyDetailsForm"
          onSubmit={(e) => {
            e.preventDefault()
            updateUserDetails(options)
              .then(() => setShowFormElement(false))
              .catch(({ graphQLErrors }) =>
                setErrorMessage(graphQLErrors[0].message)
              )
          }}
        >
          <Form.Group controlId="formPlaintextFirstName" className="pb-3">
            <Form.Label className="text-muted">First Name:</Form.Label>
            <Form.Control
              required
              type="text"
              name="firstName"
              placeholder={user.firstName}
              onChange={setInputs}
              value={inputs.firstName}
              readOnly={!showFormElement}
              isInvalid={!!errorMessage}
            />
          </Form.Group>
          <Form.Group controlId="formPlaintextPassword">
            <Form.Label className="text-muted">Last Name:</Form.Label>
            <Form.Control
              required
              type="input"
              name="lastName"
              placeholder={user.lastName}
              value={inputs.lastName}
              onChange={setInputs}
              readOnly={!showFormElement}
              isInvalid={!!errorMessage}
            />
            <Form.Control.Feedback type="invalid">
              {errorMessage}
            </Form.Control.Feedback>
          </Form.Group>
          {showFormElement ? (
            <>
              <Button
                id="MyDetailsFormSave"
                style={{ width: 100 }}
                className="float-right"
                variant="primary"
                type="submit"
              >
                Save
              </Button>

              <Button
                id="MyDetailsFormCancel"
                style={{ width: 100 }}
                type="button"
                className="float-right mr-3"
                variant="secondary"
                onClick={() => setShowFormElement(false)}
              >
                Cancel
              </Button>
            </>
          ) : (
            <Button
              id="MyDetailsFormEdit"
              style={{ width: 100 }}
              className="float-right"
              variant="primary"
              onClick={() => setShowFormElement(true)}
            >
              Edit
            </Button>
          )}
        </Form>
      </Box>
    </Row>
  )
}
