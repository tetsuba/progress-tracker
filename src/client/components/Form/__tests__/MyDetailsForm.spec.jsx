import { testRenderer } from '../../../../test/testHelper'

// COMPONENTS
import MyDetailsForm from '../MyDetailsForm'

describe('<MyDetailsForm>', () => {
  const baseProps = {
    handleSubmit: jest.fn(),
    user: {
      firstName: 'Bob',
      lastName: 'Doh',
    },
  }

  describe('@Render', () => {
    it('should render default state which is not available to edit', () => {
      // Input forms will be read only
      // Edit button will be visible
      const wrapper = testRenderer(MyDetailsForm, baseProps)
      expect(wrapper).toMatchSnapshot()
    })

    describe('User clicks on "Edit" button', () => {
      it('should render form available to edit', () => {
        // Input forms will not be read only
        // Edit button will be hidden
        // Cancel and Save buttons will be visible
        const wrapper = testRenderer(MyDetailsForm, baseProps)
        wrapper.find('#MyDetailsFormEdit').simulate('click')
        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  describe('@Events', () => {
    describe('onChange', () => {
      it('should update firstName input', () => {
        const wrapper = testRenderer(MyDetailsForm, baseProps)
        const firstName = { name: 'firstName' }
        const expected = 'Bob'

        wrapper.find(firstName).simulate('change', {
          persist: jest.fn(),
          target: {
            name: 'firstName',
            value: 'Bob',
          },
        })

        expect(wrapper.find(firstName).prop('value')).toEqual(expected)
      })

      it('should update lastName input', () => {
        const wrapper = testRenderer(MyDetailsForm, baseProps)
        const password = { name: 'lastName' }
        const expected = 'Doh'

        wrapper.find(password).simulate('change', {
          persist: jest.fn(),
          target: {
            name: 'lastName',
            value: 'Doh',
          },
        })
        expect(wrapper.find(password).prop('value')).toEqual(expected)
      })
    })

    describe('onSubmit', () => {
      it('should trigger handleSubmit', () => {
        const wrapper = testRenderer(MyDetailsForm, baseProps)
        const id = '#MyDetailsForm'
        wrapper.find(id).props().onSubmit({ preventDefault: jest.fn() })
        expect(baseProps.handleSubmit).toHaveBeenCalledTimes(1)
      })
    })
  })
})
