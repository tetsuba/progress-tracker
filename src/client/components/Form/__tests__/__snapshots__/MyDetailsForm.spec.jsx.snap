// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyDetailsForm> @Render User clicks on "Edit" button should render form available to edit 1`] = `
<MyDetailsForm
  handleSubmit={[MockFunction]}
  user={
    Object {
      "firstName": "Bob",
      "lastName": "Doh",
    }
  }
>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Box
        max={500}
      >
        <div
          style={
            Object {
              "backgroundColor": "#f8f9fa",
              "border": "solid 1px #ccc",
              "borderRadius": "8px",
              "margin": "0 auto",
              "maxWidth": 500,
              "padding": "20px 20px",
              "width": "100%",
            }
          }
        >
          <h3
            className="mb-5"
          >
            My Details
          </h3>
          <Form
            id="MyDetailsForm"
            inline={false}
            onSubmit={[Function]}
          >
            <form
              className=""
              id="MyDetailsForm"
              onSubmit={[Function]}
            >
              <FormGroup
                className="pb-3"
                controlId="formPlaintextFirstName"
              >
                <div
                  className="pb-3 form-group"
                >
                  <FormLabel
                    className="text-muted"
                    column={false}
                    srOnly={false}
                  >
                    <label
                      className="text-muted form-label"
                      htmlFor="formPlaintextFirstName"
                    >
                      First Name:
                    </label>
                  </FormLabel>
                  <FormControl
                    isInvalid={false}
                    name="firstName"
                    onChange={[Function]}
                    placeholder="Bob"
                    readOnly={true}
                    required={true}
                    type="text"
                    value="Bob"
                  >
                    <input
                      className="form-control"
                      id="formPlaintextFirstName"
                      name="firstName"
                      onChange={[Function]}
                      placeholder="Bob"
                      readOnly={true}
                      required={true}
                      type="text"
                      value="Bob"
                    />
                  </FormControl>
                </div>
              </FormGroup>
              <FormGroup
                controlId="formPlaintextPassword"
              >
                <div
                  className="form-group"
                >
                  <FormLabel
                    className="text-muted"
                    column={false}
                    srOnly={false}
                  >
                    <label
                      className="text-muted form-label"
                      htmlFor="formPlaintextPassword"
                    >
                      Last Name:
                    </label>
                  </FormLabel>
                  <FormControl
                    isInvalid={false}
                    name="lastName"
                    onChange={[Function]}
                    placeholder="Doh"
                    readOnly={true}
                    required={true}
                    type="input"
                    value="Doh"
                  >
                    <input
                      className="form-control"
                      id="formPlaintextPassword"
                      name="lastName"
                      onChange={[Function]}
                      placeholder="Doh"
                      readOnly={true}
                      required={true}
                      type="input"
                      value="Doh"
                    />
                  </FormControl>
                  <Feedback
                    type="invalid"
                  >
                    <div
                      className="invalid-feedback"
                    />
                  </Feedback>
                </div>
              </FormGroup>
              <Button
                active={false}
                className="float-right"
                disabled={false}
                id="MyDetailsFormEdit"
                onClick={[Function]}
                style={
                  Object {
                    "width": 100,
                  }
                }
                type="button"
                variant="primary"
              >
                <button
                  className="float-right btn btn-primary"
                  disabled={false}
                  id="MyDetailsFormEdit"
                  onClick={[Function]}
                  style={
                    Object {
                      "width": 100,
                    }
                  }
                  type="button"
                >
                  Edit
                </button>
              </Button>
            </form>
          </Form>
        </div>
      </Box>
    </div>
  </Row>
</MyDetailsForm>
`;

exports[`<MyDetailsForm> @Render should render default state which is not available to edit 1`] = `
<MyDetailsForm
  handleSubmit={[MockFunction]}
  user={
    Object {
      "firstName": "Bob",
      "lastName": "Doh",
    }
  }
>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Box
        max={500}
      >
        <div
          style={
            Object {
              "backgroundColor": "#f8f9fa",
              "border": "solid 1px #ccc",
              "borderRadius": "8px",
              "margin": "0 auto",
              "maxWidth": 500,
              "padding": "20px 20px",
              "width": "100%",
            }
          }
        >
          <h3
            className="mb-5"
          >
            My Details
          </h3>
          <Form
            id="MyDetailsForm"
            inline={false}
            onSubmit={[Function]}
          >
            <form
              className=""
              id="MyDetailsForm"
              onSubmit={[Function]}
            >
              <FormGroup
                className="pb-3"
                controlId="formPlaintextFirstName"
              >
                <div
                  className="pb-3 form-group"
                >
                  <FormLabel
                    className="text-muted"
                    column={false}
                    srOnly={false}
                  >
                    <label
                      className="text-muted form-label"
                      htmlFor="formPlaintextFirstName"
                    >
                      First Name:
                    </label>
                  </FormLabel>
                  <FormControl
                    isInvalid={false}
                    name="firstName"
                    onChange={[Function]}
                    placeholder="Bob"
                    readOnly={true}
                    required={true}
                    type="text"
                    value="Bob"
                  >
                    <input
                      className="form-control"
                      id="formPlaintextFirstName"
                      name="firstName"
                      onChange={[Function]}
                      placeholder="Bob"
                      readOnly={true}
                      required={true}
                      type="text"
                      value="Bob"
                    />
                  </FormControl>
                </div>
              </FormGroup>
              <FormGroup
                controlId="formPlaintextPassword"
              >
                <div
                  className="form-group"
                >
                  <FormLabel
                    className="text-muted"
                    column={false}
                    srOnly={false}
                  >
                    <label
                      className="text-muted form-label"
                      htmlFor="formPlaintextPassword"
                    >
                      Last Name:
                    </label>
                  </FormLabel>
                  <FormControl
                    isInvalid={false}
                    name="lastName"
                    onChange={[Function]}
                    placeholder="Doh"
                    readOnly={true}
                    required={true}
                    type="input"
                    value="Doh"
                  >
                    <input
                      className="form-control"
                      id="formPlaintextPassword"
                      name="lastName"
                      onChange={[Function]}
                      placeholder="Doh"
                      readOnly={true}
                      required={true}
                      type="input"
                      value="Doh"
                    />
                  </FormControl>
                  <Feedback
                    type="invalid"
                  >
                    <div
                      className="invalid-feedback"
                    />
                  </Feedback>
                </div>
              </FormGroup>
              <Button
                active={false}
                className="float-right"
                disabled={false}
                id="MyDetailsFormEdit"
                onClick={[Function]}
                style={
                  Object {
                    "width": 100,
                  }
                }
                type="button"
                variant="primary"
              >
                <button
                  className="float-right btn btn-primary"
                  disabled={false}
                  id="MyDetailsFormEdit"
                  onClick={[Function]}
                  style={
                    Object {
                      "width": 100,
                    }
                  }
                  type="button"
                >
                  Edit
                </button>
              </Button>
            </form>
          </Form>
        </div>
      </Box>
    </div>
  </Row>
</MyDetailsForm>
`;
